cleverhans/attacks.py

```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
+                              Smaller values produce better results but are
+                              slower to converge.
+        :param binary_search_steps: The number of times we perform binary
+                                    search to find the optimal tradeoff-
+                                    constant between distance and confidence.
+        :param max_iterations: The maximum number of iterations. Larger values
+                               are more accurate; setting too small will
+                               require a large learning rate and will produce
+                               poor results.
+        :param abort_early: If true, allows early aborts if gradient descent
+                            gets stuck.
+        :param initial_const: The initial tradeoff-constant to use to tune the
+                              relative importance of distance and confidence.
+                              If binary_search_steps is large, the initial
+                              constant is not important.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        import tensorflow as tf
+        from .attacks_tf import CarliniWagnerL2 as CWL2
+        self.parse_params(**kwargs)
+
+        attack = CWL2(self.sess, self.model, self.batch_size,
+                      self.confidence, self.targeted, self.learning_rate,
+                      self.binary_search_steps, self.max_iterations,
+                      self.abort_early, self.initial_const,
+                      self.clip_min, self.clip_max, self.nb_classes,
+                      x.get_shape().as_list()[1:])
+
+        if 'y' in kwargs:
+            labels = kwargs['y']
+        else:
+            if self.targeted:
+                raise ValueError("Must supply target labels 'y'"
+                                 " in targeted attack.")
+            # TODO abstract this out for other classes too
+            preds = self.model(x)
+            preds_max = tf.reduce_max(preds, 1, keep_dims=True)
+            original_predictions = tf.to_float(tf.equal(preds,
+                                                        preds_max))
+            labels = original_predictions
+
+        def cw_wrap(x_val, y_val):
+            return np.array(attack.attack(x_val, y_val), dtype=np.float32)
+        wrap = tf.py_func(cw_wrap, [x, labels], tf.float32)
+
+        return wrap
+
+    def parse_params(self, y=None, y_val=None, nb_classes=10,
+                     batch_size=1, confidence=0,
+                     targeted=True, learning_rate=5e-3,
+                     binary_search_steps=5, max_iterations=1000,
+                     abort_early=True, initial_const=1e-2,
+                     clip_min=0, clip_max=1):
+
+        # ignore the y and y_val argument
```

Comment:
Because I specifically handle them in the generate() method, and don't do anything with it in the parse_params. I might be able to restructure the code to do it in parse_params() now that generate_np() is implemented in the Attack class.